canonicalwebteam/search/views.py

```diff
@@ -45,6 +51,15 @@ def search_view():
         """
         Get search results from Google Custom Search
         """
+        # Rate limit requests to protect from spamming
+        # To adjust this rate visit
+        # https://limits.readthedocs.io/en/latest/quickstart.html#examples
+        one_per_minute = parse(request_limit)
+        rate_limit = fixed_window.hit(one_per_minute)
+        if not rate_limit:
+            return flask.abort(
+                429, "You are not allowed to request this page anymore"
```

Comment:
Feel free to suggest better descriptions.

Reply:
I think we should mention "quota" or "limit", and I also have no problem mentioning the amount explicitly. How about:

"Sorry, you're only allowed to perform up to 500 searches per day. Please try again tomorrow. Please [reach out to us](https://github.com/canonical/ubuntu.com/issues/new) if you feel this limit should be changed."

Reply:
Although, how much of that should be on search and how much on the parent project? Maybe the "reach out to us" part could be left up to the site to add. So just:

> Sorry, you're only allowed to perform up to 500 searches per day. Please try again tomorrow.

Reply:
Well, usually they suggest adding the rate limits in the headers. Specifically in the case of search, I thought we shouldn't show any specific numbers, as to avoid spammers getting a hint on how to circumvent the limit.

Can we do it without specifying the limit number? Whether the limit should be changed is should be up to us? Something like "You are allowed to perform up to a limited number of searches per day, please try again tomorrow"?

Reply:
Yeah, and each project can actually override this. So I think this should be as standard as possible, but easy to understand so that each project knows what this is about.

Reply:
> Well, usually they suggest adding the rate limits in the headers.

What do you mean? Nothing here adds anything to the headers does it?

> Specifically in the case of search, I thought we shouldn't show any specific numbers, as to avoid spammers getting a hint on how to circumvent the limit.

This is basically security through obscurity. In this case, it's fairly trivial for them to find out the rate limit through experimentation, and encouraging such experimentation would obviously add more load to our API. The reason I want to communicate clearly about this is because we really don't know if we're going to be hurting real people with real honest reasons for wanting to use our search. And I don't understand at all why you wouldn't want them to reach out to us? We need as much information about this as we can get.

> Yeah, and each project can actually override this.

That's a good point. So we should simply print the rate limit out in the message.

```
f"The rate limit is: {rate_limit}"
```

Reply:
Yes, they can find out about the rate limit through experimentation, but they can't automate it. If we output the `rate_limit` somewhere, it'd be quite straightfoward to set up something to avoid hitting rate limit and then keep spamming

Reply:
Alright let's see if that happens.

Reply:
It's trivial to write a script that responds to 429 by backing off a bit and then trying again a bit later. I've done so myself for the Discourse API - who tell you the exact number of seconds you have to wait, but even if they didn't, just having a bounce in my script would be almost as effective.

There's really no point trying to protect ourselves by obscuring stuff. We should protect ourselves by setting the right limits to protect ourselves.

Reply:
[Security through obscurity](https://en.wikipedia.org/wiki/Security_through_obscurity) is a well-known antipattern. NIST are perhaps the most high-profile body to [explicitly advise](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-123.pdf) open design for security:

> Open Design—System security should not depend on the secrecy of the implementation or its components.

Reply:
Updated